Design/name-mangling.md

```diff
@@ -78,111 +89,149 @@ We can identify the possible local names in this example:
 * Setter of `var x` in `B` -> `prop_x_setter`, `prop_x`, `x`
 * Parameter `x` in `fun foo(...)`: `x`
 
-## Approach
-
-We can regard the name mangling problem as a series of
-algorithms of the following form:
-
-1. We have a set of objects `X`.
-2. Each `x: X` has a list of preferred names `x.names`,
-   where each preferred name is marked primary or secondary.
-  * We are guaranteed that all primary names are unique.
-  * Secondary names may clash.
-3. We create a map from names to objects, indicating which
-   objects laid a claim to what name.
-4. We resolve the claims.  For each `x: X`:
-  * If `x` has a secondary name claimed by no other
-    object, it gets that name.
-  * Otherwise, `x` gets its primary name.
-  * Note that if two objects claimed the same name as
-    primary, the uniqueness invariant was violated, so the
-    algorithm fails.
-
-We run this algorithm repeatedly:
-1. For every namespace, we resolve local names.
-2. We resolve qualified names that appear in namespaces.
-3. We resolve the remaining qualified names.
-
-### Annotated example
-
-Let us return to the example above, with the preferred names
-annotated:
-```kotlin
-// A: A, class_A
-// x: x, field_x
-class A(val x: Int) {
-    // foo: foo, fun_foo, fun_foo_takes_Unit, fun_foo_takes_Unit_returns_Int
-    fun foo(): Int = ...
-}
+## Algorithm
 
-// B: class_B, B
-class B {
-    // get x: x, prop_x, prop_x_getter
-    // set x: x, prop_x, prop_x_setter
-    var x: Int
-        get() { ... }
-        set(v) { ... }
-}
+We work with two graphs:
 
-// A: A, fun_A, fun_A_takes_Unit, fun_A_takes_Unit_returns_Unit
-fun A() { ... }
+a **dependency graph** whose vertices are pairs 
+$(x, \text{version})$,  where $x$ is a mangled name 
+and $\text{version} \in \{\text{short}, \text{medium}, 
+\text{long}, \ldots\}$.  
+Edges encode that rendering $(x, v_x)$ requires the representation 
+of $(y, v_y)$.
```

Comment:
But does `(x, v_x)` know which representation it needs?  Like, doesn't it just depend on `y`?